OpenRA.Game/Graphics/Util.cs

```diff
@@ -93,6 +93,69 @@ public static class Util
 			vertices[nv + 5] = new Vertex(a, r.Left, r.Top, sl, st, paletteTextureIndex, fAttribC, tint, alpha);
 		}
 
+		public static void FastCreateQuad4Vertex(Vertex[] vertices, in float3 o, Sprite r, int2 samplers, float paletteTextureIndex, int nv,
+			in float3 size, in float3 tint, float alpha, float rotation = 0f)
+		{
+			float3 a, b, c, d;
+
+			// Rotate sprite if rotation angle is not equal to 0
+			if (rotation != 0f)
+			{
+				var center = o + 0.5f * size;
+				var angleSin = (float)Math.Sin(-rotation);
+				var angleCos = (float)Math.Cos(-rotation);
+
+				// Rotated offset for +/- x with +/- y
+				var ra = 0.5f * new float3(
+					size.X * angleCos - size.Y * angleSin,
+					size.X * angleSin + size.Y * angleCos,
+					(size.X * angleSin + size.Y * angleCos) * size.Z / size.Y);
+
+				// Rotated offset for +/- x with -/+ y
+				var rb = 0.5f * new float3(
+					size.X * angleCos + size.Y * angleSin,
+					size.X * angleSin - size.Y * angleCos,
+					(size.X * angleSin - size.Y * angleCos) * size.Z / size.Y);
+
+				a = center - ra;
+				b = center + rb;
+				c = center + ra;
+				d = center - rb;
+			}
+			else
+			{
+				a = o;
+				b = new float3(o.X + size.X, o.Y, o.Z);
+				c = new float3(o.X + size.X, o.Y + size.Y, o.Z + size.Z);
+				d = new float3(o.X, o.Y + size.Y, o.Z + size.Z);
+			}
+
+			float sl = 0;
+			float st = 0;
+			float sr = 0;
+			float sb = 0;
+
+			// See combined.vert for documentation on the channel attribute format
+			var attribC = r.Channel == TextureChannel.RGBA ? 0x02 : ((byte)r.Channel) << 1 | 0x01;
+			attribC |= samplers.X << 6;
+			if (r is SpriteWithSecondaryData ss)
+			{
+				sl = ss.SecondaryLeft;
+				st = ss.SecondaryTop;
+				sr = ss.SecondaryRight;
+				sb = ss.SecondaryBottom;
+
+				attribC |= ((byte)ss.SecondaryChannel) << 4 | 0x08;
+				attribC |= samplers.Y << 9;
+			}
+
+			var fAttribC = (float)attribC;
+			vertices[nv] = new Vertex(a, r.Left, r.Top, sl, st, paletteTextureIndex, fAttribC, tint, alpha);
+			vertices[nv + 1] = new Vertex(b, r.Right, r.Top, sr, st, paletteTextureIndex, fAttribC, tint, alpha);
+			vertices[nv + 2] = new Vertex(c, r.Right, r.Bottom, sr, sb, paletteTextureIndex, fAttribC, tint, alpha);
+			vertices[nv + 3] = new Vertex(d, r.Left, r.Bottom, sl, sb, paletteTextureIndex, fAttribC, tint, alpha);
```

Comment:
Is this method copy-paste aside from assigning only 4 verticies at the end? Extract the common code to a helper method.

Reply:
I plan to replace the above functions fairly soon given how massive of a performance gain I see. Would this be temporarily duplicated be acceptable?

Reply:
I'd prefer to see it changed. It'll take 2 minutes now and be nicer when inevitably "soon" turns out to be longer than expected depending on review speed, etc.

Reply:
ok

Reply:
removed the redundant changes as that "soon" did come